### PR TITLE
Gateway API Translator support for GRPCRoute

### PR DESCRIPTION
**Pull request by arkodg**
Relates to https://github.com/envoyproxy/gateway/issues/642

Signed-off-by: Arko Dasgupta <arko@tetrate.io>

### CODE REVIEW

**Comment by quyenhoang96** [comment]
Please finish this PR. i think it is an important PR

File: internal/gatewayapi/route.go [hunk 1]
```diff
@@ -50,6 +51,31 @@ func (t *Translator) ProcessHTTPRoutes(httpRoutes []*v1beta1.HTTPRoute, gateways
 	return relevantHTTPRoutes
 }
 
+func (t *Translator) ProcessGRPCRoutes(grpcRoutes []*v1alpha2.GRPCRoute, gateways []*GatewayContext, resources *Resources, xdsIR XdsIRMap) []*GRPCRouteContext {
+	var relevantGRPCRoutes []*GRPCRouteContext
+
+	for _, g := range grpcRoutes {
+		if g == nil {
+			panic("received nil httproute")
```
Comment (danehans) [inline internal/gatewayapi/route.go:59]: s/httproute/grpcroute/

File: internal/gatewayapi/route.go [hunk 2]
```diff
@@ -251,11 +277,147 @@ func applyHTTPFiltersContexttoIRRoute(httpFiltersContext *HTTPFiltersContext, ir
...
+		// If the route has no valid backends then just use a direct response and don't fuss with weighted responses
+		for _, ruleRoute := range ruleRoutes {
+			if ruleRoute.BackendWeights.Invalid > 0 && len(ruleRoute.Destinations) == 0 {
+				ruleRoute.DirectResponse = &ir.DirectResponse{
+					StatusCode: 500,
+				}
+			}
+		}
+
+		// TODO handle:
```
Comment (danehans) [inline internal/gatewayapi/route.go:352]: Do we have an issue for this TODO? If not, please create one.

File: internal/gatewayapi/route.go [hunk 3]
```diff
@@ -251,11 +277,147 @@ func applyHTTPFiltersContexttoIRRoute(httpFiltersContext *HTTPFiltersContext, ir
...
+			if match.Method.Method != nil {
+				irRoute.HeaderMatches = append(irRoute.HeaderMatches, &ir.StringMatch{
+					Name:  ":method",
+					Exact: match.Method.Method,
+				})
+			}
+			/* TODO
+			if match.Method.Service != nil {
+			}
+			*/
```
Comment (danehans) [inline internal/gatewayapi/route.go:407]: Do we have an issue for this TODO? If not, please create one.

**Review by danehans — APPROVED** [review]
@arkodg my feedback is generally nits that can be resolved in a follow-on PR so I have approved.

File: internal/gatewayapi/1 [hunk 4]
```diff
@@ -0,0 +1,95 @@
+gateways:
```
[the remark this answers is not in the document]
Comment (arkodg) [reply internal/gatewayapi/1]: its an outdated file, not part of the final commit

File: internal/gatewayapi/route.go (hunk 1 above)
Comment (arkodg) [reply internal/gatewayapi/route.go:59]: fixed in https://github.com/envoyproxy/gateway/pull/962

File: internal/gatewayapi/filters.go [hunk 5]
```diff
@@ -78,36 +78,74 @@ func (t *Translator) ProcessHTTPFilters(parentRef *RouteParentContext,
...
+			t.processExtensionRefHTTPFilter(filter.ExtensionRef, httpFiltersContext, resources)
+		default:
+			t.processUnsupportedHTTPFilter(string(filter.Type), httpFiltersContext)
+		}
+	}
+
+	return httpFiltersContext
+}
+
+// ProcessGRPCFilters translate gateway api grpc filters to IRs.
```
[the remark this answers is not in the document]
Comment (arkodg) [reply internal/gatewayapi/filters.go:106]: https://github.com/envoyproxy/gateway/pull/962

File: internal/gatewayapi/contexts.go [hunk 6]
```diff
@@ -243,6 +248,91 @@ func (h *HTTPRouteContext) GetRouteParentContext(forParentRef v1beta1.ParentRefe
 	return ctx
 }
 
+// GRPCRouteContext wraps an GRPCRoute and provides helper methods for
```
[the remark this answers is not in the document]
Comment (arkodg) [reply internal/gatewayapi/contexts.go:253]: fixed in https://github.com/envoyproxy/gateway/pull/962

**Pull request merged**